### PR TITLE
[FIX] website: correctly toggle the preview for the countdown snippet

### PR DESCRIPTION
**Pull request by tdecaluwe**
The countdown widget has an end action that can be configured to show a message when the countdown reaches zero. A button in the editor toggles a preview of this message. However, a bug currently makes the preview disappear whenever the widget is restarted. Such a restart is triggered whenever the editor emits a `widgets_start_request` event is emitted in the editor.

To solve the problem, the preview visibility is now controlled by a separate css class `s_countdown_enable_preview` overriding `d-none`. This way, the preview visibility no longer interacts with the widget's logic and is no longer affected by the widget restarting.

A second commit avoids content below a countdown snippet from jumping (perceived as a flicker) by setting a `min-height` on the countdown block in the html (so it is already there before it is rendered). 

task-2638366

### CODE REVIEW

**Comment by rdeodoo** [comment]
Task number?
Wrong version, bug also in 14.0

**Comment by rdeodoo** [comment]
Also it changes the behavior -> The countdown should always be shown, only the message should be hidden/shown on click

**Comment by tdecaluwe** [comment]
@rdeodoo, ok, I will retarget the PR

File: addons/website/static/src/snippets/s_countdown/000.scss [hunk 1]
```diff
@@ -0,0 +1,7 @@
+// s_countdown
+.s_countdown_enable_preview {
+    display: initial !important;
+}
+.s_countdown_none {
+    display: none !important;
+}
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_countdown/000.scss:7]: This should not be in a 000.scss file, those are dedicated to style the snippet for the visitors and support snippet versioning. Those classes here are edit mode classes, that the visitors should thus not have to load. They also do not need versionning as our options are only meant to work on the last version of snippets.

You can thus put those in website.edit_mode.scss, no need of a new file. Besides, even if that was the correct way to go, your diff would not be ok as it is not a stable fix: you modify the XML assets, meaning users would have to do a -u, which they don't in stable versions. So, here someone getting your fix would only receive the JS part, toggling classes which have no effect (you would thus worsen the bug).

File: addons/website/static/src/snippets/s_countdown/options.js [hunk 2]
```diff
@@ -12,6 +12,11 @@ options.registry.countdown = options.Class.extend({
         'click .toggle-edit-message': '_onToggleEndMessageClick',
     }),
 
+    cleanForSave: async function () {
+        this.$target.find('.s_countdown_canvas_wrapper').removeClass("s_countdown_none");
+        this.$target.find('.s_countdown_end_message').removeClass("s_countdown_enable_preview");
+        this._super(...arguments);
+    },
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_countdown/options.js:19]: - Missing blank line before options section, missing documentation markers
- No need of calling super in cleanForSave, it's meant to be forever abstract in the parent definition

File: addons/website/static/src/snippets/s_countdown/options.js [hunk 3]
```diff
@@ -28,6 +33,8 @@ options.registry.countdown = options.Class.extend({
                 const message = this.endMessage || qweb.render('website.s_countdown.end_message');
                 this.$target.append(message);
             }
+            this.$target.toggleClass('hide-countdown', widgetValue === 'message_no_countdown');
+            this.updateUIEndMessage();
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_countdown/options.js:37]: What's that? Does not seem related to the fix

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website/static/src/snippets/s_countdown/options.js (hunk 3 above)
Comment (tdecaluwe) [reply addons/website/static/src/snippets/s_countdown/options.js:37]: Yes, this remains from the original fix targeting `saas-14.5` by mistake. It should be removed when targeting `14.0`, `message_no_countdown` does not yet exist in this version.

Comment (rdeodoo) [reply addons/website/static/src/snippets/s_countdown/options.js]: Just to be sure, what did you chose? 2 PR or add second fix in forward port?
Don't forget to add to `up to <version>` if this is not master

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@tdecaluwe Can you squash what needs to be squashed? Thank you :+1:

File: addons/website/static/src/snippets/s_countdown/options.js (hunk 3 above)
Comment (rdeodoo) [reply addons/website/static/src/snippets/s_countdown/options.js]: J'imagine que l'avenir nous le dira :shrug:

**Comment by rdeodoo** [comment]
https://github.com/odoo/odoo/pull/76343#issuecomment-918195819

> Task number?

image

:shrug:

**Comment by tdecaluwe** [comment]
@rdeodoo I added it in the PR description before, I forgot the commits... 🙃

**Comment by rdeodoo** [comment]
Don't forget to `up to <version>` here so you can introduced the second fix in 14.4/14.5 (I don't remember where it was when we discussed it).
I didn't see any second PR link on the task so I guess you kept the second fix for this PR forward port as discussed (https://github.com/odoo/odoo/pull/76343#discussion_r708433641).

File: addons/website/static/src/snippets/s_countdown/options.js (hunk 3 above)
Comment (tdecaluwe) [reply addons/website/static/src/snippets/s_countdown/options.js]: I'll add the second fix in the forward port

**Comment by tdecaluwe** [comment]
@rdeodoo Yes the second fix is for the forward port.

**Comment by tdecaluwe** [comment]
@rdeodoo up to version 14.4 (the second bug was introduced in 14.4)

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Some comments. In fact kinda debatable that this fix lands in 14.0 but I guess it's ok. However, the original task's spec is in fact clearly not related and you can already create another PR targeting the right version to fix it (I think we decided to do it in the forward-port but it in fact does not make much sense... or can you remind me why?). Thanks.

**Review by bvr-odoo — CHANGES_REQUESTED** [review]
Maybe update the PR description if you fix the bug described in the associated task in another PR and link it in the task with an explanation. Because it's not clear, where is the fix for the message that disappears?

**Comment by tdecaluwe** [comment]
@bvr-odoo A commit went missing... The PR should be good now (I'll add a note about the fix for the flicker)

**Review by qsm-odoo — COMMENTED** [review]
I will push the updates needed myself in a few seconds. For the record, here are some comments.

File: addons/website/static/tests/tours/snippet_countdown.js [hunk 4]
```diff
@@ -33,5 +33,14 @@ tour.register('snippet_countdown', {
...
     },
+    // Next, we change the end action to redirect while the edit message toggle
+    // is still activated. Now, we can check if the countdown wasn't
+    // incorrectly hidden.
+    wTourUtils.changeOption('countdown', 'we-title:contains("at the end") + div we-toggler'),
+    wTourUtils.changeOption('countdown', 'we-button:contains("Redirect")', 'end action'),
+    {
+        trigger: '.s_countdown_canvas_wrapper',
+        run: () => null,
+    },
```
[the remark this answers is not in the document]
Comment (tdecaluwe) [reply addons/website/static/tests/tours/snippet_countdown.js]: @qsm-odoo I double checked and this part of the test does fail when I comment the lines from the same commit.

**Review by qsm-odoo — APPROVED** [review]
@robodoo rebase-ff r+

File: addons/website/static/src/snippets/s_countdown/options.js (hunk 3 above)
Comment (rdeodoo) [reply addons/website/static/src/snippets/s_countdown/options.js]: Do you already have that second fix? (The second issue we discussed a loooooooooooot of time?)
If yes, maybe provide it to @qsm-odoo which just r+ the PR and will handle it

File: addons/website/static/tests/tours/snippet_countdown.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/website/static/tests/tours/snippet_countdown.js]: Which precisely is disturbing as, after your first commit, following the flow of your test as a real user leads to **no bug**: https://watch.screencastify.com/v/8oB9zKtMPvb315AiMSbS since your flow was not actually *clicking* on the "hide at the end" button, so you are saying that the test revealed a bug which did not exist.

**Comment by rdeodoo** [comment]
@qsm-odoo By the way, the forward bot got disabled, on this PR, see https://github.com/odoo/odoo/pull/76343#issuecomment-930161551
It was `up to` 14.4 to insert the second fix, but since that version got dropped, it basically mean it is disabled.
